src/hotspot/share/memory/metaspace/blockTree.cpp

```diff
@@ -0,0 +1,207 @@
+/*
+ * Copyright (c) 2020, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2020 SAP SE. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "memory/metaspace/blockTree.hpp"
+#include "memory/resourceArea.hpp"
+#include "utilities/debug.hpp"
+#include "utilities/globalDefinitions.hpp"
+#include "utilities/growableArray.hpp"
+#include "utilities/ostream.hpp"
+
+namespace metaspace {
+
+// Needed to prevent linker errors on MacOS and AIX
+const size_t BlockTree::MinWordSize;
+
+#ifdef ASSERT
+
+// Tree verification
+
+// These asserts prints the tree, then asserts
+#define assrt(cond, format, ...) \
+  do { \
+    if (!(cond)) { \
+      print_tree(tty); \
+      assert(cond, format, __VA_ARGS__); \
+    } \
+  } while(0)
+
+  // This assert prints the tree, then stops (generic message)
+#define assrt0(cond) \
+  do { \
+    if (!(cond)) { \
+      print_tree(tty); \
+      assert(cond, "sanity"); \
+    } \
+  } while(0)
+
+
+
+// walkinfo keeps a node plus the size corridor it and its children
+//  are supposed to be in.
+struct BlockTree::walkinfo {
+  BlockTree::Node* n;
+  int depth;
+  size_t lim1; // (
+  size_t lim2; // )
+};
+
+void BlockTree::verify() const {
```

Comment:
> I think this sort of deep verification should be in a gtest instead, at least for blockTree. Note, this can be fixed/discussed in a future RFE.

:( This would affect a lot of my code since I do a lot of in-file verifications. Yeah, lets discuss this separately.